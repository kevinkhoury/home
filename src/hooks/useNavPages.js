import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query NavPagesQuery {
        markdownRemark(frontmatter: { templateKey: { eq: "menu-data" } }) {
          frontmatter {
            menuItems {
              slug
              label
            }
          }
        }
      }
    `,
  )
  const { menuItems } = query.markdownRemark.frontmatter
  return menuItems.map(({ slug: orginalSlug, label }) => {
    // normalize to initial and trailing slash
    const slug =
      orginalSlug === '/'
        ? orginalSlug
        : `/${orginalSlug
            .split('/')
            .filter(x => x)
            .join('/')}/`
    return { slug, label }
  })
  // return query.markdownRemark.frontmatter.menuItems
}
