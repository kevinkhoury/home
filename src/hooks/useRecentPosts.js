import { useStaticQuery, graphql } from 'gatsby'
import { getValidDates } from '../utils'

export default () => {
  const data = useStaticQuery(
    graphql`
      query QueryRecentPosts {
        allMarkdownRemark(
          sort: { order: DESC, fields: [frontmatter___date] }
          filter: {
            frontmatter: {
              templateKey: { eq: "blog-post" }
              published: { eq: true }
            }
          }
          limit: 4
        ) {
          edges {
            node {
              fields {
                slug
                gitAuthorTime
                gitCreatedTime
              }
              frontmatter {
                date(formatString: "MMM D, YYYY")
                pageTitle
                featuredImage {
                  src {
                    childImageSharp {
                      fluid {
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  m: src {
                    childImageSharp {
                      fluid(maxWidth: 500, maxHeight: 664) {
                        ...GatsbyImageSharpFluid_withWebp
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  d: src {
                    childImageSharp {
                      fluid(maxWidth: 1000, maxHeight: 664) {
                        ...GatsbyImageSharpFluid_withWebp
                        originalName
                      }
                      original {
                        height
                        width
                      }
                    }
                  }
                  alt
                }
              }
            }
          }
        }
      }
    `,
  )
  if (
    data &&
    data.allMarkdownRemark &&
    data.allMarkdownRemark.edges &&
    data.allMarkdownRemark.edges.length > 0
  ) {
    return data.allMarkdownRemark.edges.map(({ node }) => {
      const {
        frontmatter: { featuredImage, pageTitle, date: userDate },
        fields: { slug, gitAuthorTime, gitCreatedTime },
      } = node
      const { date, dateModified } = getValidDates(
        userDate,
        gitAuthorTime,
        gitCreatedTime,
      )
      return {
        image: !!featuredImage ? featuredImage : null,
        slug,
        pageTitle,
        date,
        dateModified,
      }
    })
  } else {
    return []
  }
}
