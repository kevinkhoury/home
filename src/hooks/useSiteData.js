import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query SiteDataQuery {
        markdownRemark(frontmatter: { templateKey: { eq: "site-data" } }) {
          frontmatter {
            name
            jobTitle
            siteName
            siteUrl
            socialLinks {
              twitter {
                url
                show
              }
              facebook {
                url
                show
              }
              linkedin {
                url
                show
              }
              pinterest {
                url
                show
              }
              instagram {
                url
                show
              }
            }
            themeOptions {
              colorScheme
              fontScheme
              showThemeSwitcher
            }
            fallbackImage {
              childImageSharp {
                original {
                  height
                  width
                }
                fluid(maxWidth: 1440) {
                  ...GatsbyImageSharpFluid_withWebp
                  originalName
                }
              }
            }
          }
        }
      }
    `,
  )
  return query.markdownRemark.frontmatter
}
