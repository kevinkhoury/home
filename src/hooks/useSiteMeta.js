import { useStaticQuery, graphql } from 'gatsby'

export default () => {
  const query = useStaticQuery(
    graphql`
      query QuerySiteMetadata {
        site {
          siteMetadata {
            title
            colorOptions {
              value
              label
            }
            fontOptions {
              value
              label
            }
          }
        }
      }
    `,
  )
  return query.site.siteMetadata
}
