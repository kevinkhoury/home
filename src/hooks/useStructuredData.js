const buildImage = (image) => ({
  '@context': 'http://schema.org/',
  '@type': 'ImageObject',
  url: image.url,
  height: image.height,
  width: image.width,
})

const personSchema = ({ name, jobTitle, sameAs }) => ({
  '@context': 'http://schema.org/',
  '@type': 'Person',
  '@id': '#person',
  name,
  jobTitle,
  sameAs,
})

const publisherSchema = ({ siteName, logo }) => ({
  '@context': 'http://schema.org/',
  '@type': 'Organization',
  '@id': '#publisher',
  name: siteName,
  logo: buildImage(logo),
})

const webpageSchema = ({
  headline,
  url,
  description,
  image,
  name,
  jobTitle,
  schemaType,
  sameAs,
}) => {
  const pageSchema = {
    '@context': 'http://schema.org',
    '@type': schemaType,
    headline,
    url,
    description,
    primaryImageOfPage: buildImage(image),
    author: {
      '@id': '#person',
    },
    copyrightHolder: {
      '@id': '#person',
    },
  }

  return JSON.stringify([personSchema({ name, jobTitle, sameAs }), pageSchema])
}

const articleSchema = ({
  headline,
  url,
  description,
  image,
  fallbackImage,
  datePublished,
  dateModified,
  name,
  siteName,
  jobTitle,
  sameAs,
}) => {
  const articleSchema = {
    '@context': 'http://schema.org',
    '@type': 'BlogPosting',
    headline,
    url,
    description,
    image: buildImage(image),
    author: {
      '@id': '#person',
    },
    datePublished,
    dateModified,
    publisher: {
      '@id': '#publisher',
    },
    mainEntityOfPage: {
      '@type': 'WebPage',
      '@id': url,
    },
  }

  return JSON.stringify([
    personSchema({ name, jobTitle, sameAs }),
    publisherSchema({ siteName, logo: fallbackImage }),
    articleSchema,
  ])
}

export default function (data) {
  const isPage = data.templateKey.indexOf('-page') !== -1
  const isArticle = data.templateKey.indexOf('-post') !== -1
  const isArchive = data.templateKey.indexOf('-archive') !== -1

  if (isPage || isArchive) {
    return webpageSchema(data)
  } else if (isArticle) {
    return articleSchema(data)
  }
}
