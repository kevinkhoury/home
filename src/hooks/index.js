import useNavPages from './useNavPages'
import useSiteData from './useSiteData'
import useStructuredData from './useStructuredData'
import useSiteMeta from './useSiteMeta'
import useRecentPosts from './useRecentPosts'

export {
  useNavPages,
  useSiteData,
  useStructuredData,
  useSiteMeta,
  useRecentPosts,
}
