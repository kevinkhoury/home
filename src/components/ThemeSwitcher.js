/* eslint jsx-a11y/no-onchange: 0 */ // --> OFF

import React from 'react'
import PropTypes from 'prop-types'
import { v4 as uuidv4 } from 'uuid'
import { useSiteMeta } from '../hooks'

const ThemeSwitcher = ({
  colorScheme,
  setColorScheme,
  fontScheme,
  setFontScheme,
  previewOpen,
  setPreviewOpen,
}) => {
  const { colorOptions, fontOptions } = useSiteMeta()

  const outerStyle = {
    position: 'fixed',
    top: '80px',
    left: '20px',
    zIndex: '50',
    boxShadow: '3px 3px 5px rgba(0,0,0,0.25)',
    borderRadius: '0.5rem',
    background: '#fff',
  }

  const wrapperStyle = {
    padding: previewOpen ? '10px' : '0',
    width: 'auto',
    height: 'auto',
    maxHeight: previewOpen ? '300px' : '30px',
    maxWidth: previewOpen ? '220px' : '30px',
    transition: 'all 0.15s ease',
    overflow: 'hidden',
  }

  const buttonStyle = {
    padding: '0 5px',
    lineHeight: '30px',
    height: '30px',
    margin: '0',
    display: 'block',
    width: '30px',
    position: 'absolute',
    right: previewOpen ? '-10px' : '0',
    top: previewOpen ? '-10px' : '0',
    outline: 'none',
    boxShadow: 'none',
  }

  const togglePreview = () => {
    setPreviewOpen(!previewOpen)
  }

  return (
    <div id="theme-switcher" style={outerStyle}>
      <button onClick={togglePreview} style={buttonStyle}>
        {previewOpen ? 'Ã—' : 'P'}
      </button>
      <div style={wrapperStyle}>
        <strong>Preview Theme Options</strong>
        <div style={{ marginBottom: '0.75em' }}>
          NOTE: This does not save changes, it only previews
        </div>
        <label htmlFor="theme-preview-colors">Color Scheme</label>
        <select
          id="theme-preview-colors"
          onChange={(e) => {
            setColorScheme(e.target.value)
          }}
          value={colorScheme}
        >
          {colorOptions.map(({ value, label }) => (
            <option key={uuidv4()} value={value}>
              {label}
            </option>
          ))}
        </select>
        <label htmlFor="theme-preview-fonts">Font Scheme</label>
        <select
          id="theme-preview-fonts"
          onChange={(e) => {
            setFontScheme(e.target.value)
          }}
          value={fontScheme}
        >
          {fontOptions.map(({ value, label }) => (
            <option key={uuidv4()} value={value}>
              {label}
            </option>
          ))}
        </select>
      </div>
    </div>
  )
}

ThemeSwitcher.propTypes = {
  colorScheme: PropTypes.string.isRequired,
  setColorScheme: PropTypes.func.isRequired,
  fontScheme: PropTypes.string.isRequired,
  setFontScheme: PropTypes.func.isRequired,
  previewOpen: PropTypes.bool.isRequired,
  setPreviewOpen: PropTypes.func.isRequired,
}

export default ThemeSwitcher
