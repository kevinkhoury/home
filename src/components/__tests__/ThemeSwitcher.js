import React from 'react'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import ThemeSwitcher from '../ThemeSwitcher'
import userEvent from '@testing-library/user-event'
import { useSiteMeta as useSiteMetaMock } from '../../hooks'

const DEFAULT_PROPS = {
  colorScheme: 'green',
  setColorScheme: jest.fn(),
  fontScheme: 'roboto',
  setFontScheme: jest.fn(),
  previewOpen: true,
  setPreviewOpen: jest.fn(),
}

jest.mock('../../hooks', () => ({
  useSiteMeta: jest.fn(),
}))

jest.mock('uuid', () => ({
  v4: jest
    .fn(() => {
      const { v4 } = jest.requireActual('uuid')
      return v4()
    })
    .mockImplementationOnce(() => 'uuid1')
    .mockImplementationOnce(() => 'uuid2')
    .mockImplementationOnce(() => 'uuid3')
    .mockImplementationOnce(() => 'uuid4'),
}))

beforeEach(() => {
  useSiteMetaMock.mockImplementation(() => ({
    colorOptions: [
      { value: 'green', label: 'green' },
      { value: 'red', label: 'red' },
    ],
    fontOptions: [
      { value: 'roboto', label: 'roboto' },
      { value: 'arial', label: 'arial' },
    ],
  }))
})

afterEach(() => {
  jest.clearAllMocks()
})

describe('ThemeSwitcher', () => {
  it('snapshot renders properly', () => {
    const tree = renderer.create(<ThemeSwitcher {...DEFAULT_PROPS} />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('toggles theme switcher menu on button click', () => {
    render(<ThemeSwitcher {...DEFAULT_PROPS} />)
    userEvent.click(screen.getByRole('button'))
    expect(DEFAULT_PROPS.setPreviewOpen.mock.calls.length).toBe(1)
  })
  it('updates font on font selection', () => {
    render(<ThemeSwitcher {...DEFAULT_PROPS} />)
    userEvent.selectOptions(screen.getByLabelText('Font Scheme'), 'arial')
    expect(DEFAULT_PROPS.setFontScheme.mock.calls.length).toBe(1)
  })
  it('updates color on color selection', () => {
    render(<ThemeSwitcher {...DEFAULT_PROPS} />)
    userEvent.selectOptions(screen.getByLabelText('Color Scheme'), 'red')
    expect(DEFAULT_PROPS.setColorScheme.mock.calls.length).toBe(1)
  })
})
