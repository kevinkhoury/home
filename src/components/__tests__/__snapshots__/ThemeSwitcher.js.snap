// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher snapshot renders properly 1`] = `
<div
  id="theme-switcher"
  style={
    Object {
      "background": "#fff",
      "borderRadius": "0.5rem",
      "boxShadow": "3px 3px 5px rgba(0,0,0,0.25)",
      "left": "20px",
      "position": "fixed",
      "top": "80px",
      "zIndex": "50",
    }
  }
>
  <button
    onClick={[Function]}
    style={
      Object {
        "boxShadow": "none",
        "display": "block",
        "height": "30px",
        "lineHeight": "30px",
        "margin": "0",
        "outline": "none",
        "padding": "0 5px",
        "position": "absolute",
        "right": "-10px",
        "top": "-10px",
        "width": "30px",
      }
    }
  >
    Ã—
  </button>
  <div
    style={
      Object {
        "height": "auto",
        "maxHeight": "300px",
        "maxWidth": "220px",
        "overflow": "hidden",
        "padding": "10px",
        "transition": "all 0.15s ease",
        "width": "auto",
      }
    }
  >
    <strong>
      Preview Theme Options
    </strong>
    <div
      style={
        Object {
          "marginBottom": "0.75em",
        }
      }
    >
      NOTE: This does not save changes, it only previews
    </div>
    <label
      htmlFor="theme-preview-colors"
    >
      Color Scheme
    </label>
    <select
      id="theme-preview-colors"
      onChange={[Function]}
      value="green"
    >
      <option
        value="green"
      >
        green
      </option>
      <option
        value="red"
      >
        red
      </option>
    </select>
    <label
      htmlFor="theme-preview-fonts"
    >
      Font Scheme
    </label>
    <select
      id="theme-preview-fonts"
      onChange={[Function]}
      value="roboto"
    >
      <option
        value="roboto"
      >
        roboto
      </option>
      <option
        value="arial"
      >
        arial
      </option>
    </select>
  </div>
</div>
`;
